geo/src/algorithm/coords_iter.rs

```diff
@@ -95,476 +99,830 @@ pub trait CoordsIter {
+                .sum::<usize>(),
+            GeometryType::MultiPolygon(mp) => MultiPolygonTrait::polygons(mp)
+                .map(|p| polygon_coords_count(&p))
+                .sum::<usize>(), // Pass ref to helper
+            GeometryType::GeometryCollection(gc) => GeometryCollectionTrait::geometries(gc)
+                .map(|g| g.coords_count())
+                .sum::<usize>(),
+            GeometryType::Rect(_r) => 5,     // 4 corners + closing coord
+            GeometryType::Triangle(_t) => 4, // 3 corners + closing coord
+        }
     }
 
-    fn exterior_coords_iter(&self) -> Self::ExteriorIter<'_> {
-        self.coords_iter()
+    fn exterior_coords_iter(&self) -> ExteriorIter<G> {
+        match self.as_type() {
+            GeometryType::Point(p) => ExteriorIter::Point(Some(PointTrait::coord(p).unwrap().to_coord())),
+            GeometryType::Line(l) => ExteriorIter::Line(LineIter { line: l, index: 0 }),
+            GeometryType::LineString(ls) => ExteriorIter::LineString(LineStringIter {
+                linestring: ls,
+                index: 0,
+            }),
+            GeometryType::Polygon(p) => ExteriorIter::Polygon(PolygonExteriorIter {
+                polygon: p,
+                index: 0,
+            }),
+            GeometryType::MultiPoint(mp) => ExteriorIter::MultiPoint(MultiPointIter {
+                multi_point: mp,
+                index: 0,
+            }),
+            GeometryType::MultiLineString(mls) => {
+                ExteriorIter::MultiLineString(MultiLineStringIter::new(mls))
+            }
+            GeometryType::MultiPolygon(mp) => {
+                ExteriorIter::MultiPolygon(MultiPolygonExteriorIter::new(mp))
+            }
+            GeometryType::GeometryCollection(gc) => {
+                // For GeometryCollection, we need to collect into a Vec first
+                // because GeometryTrait is not dyn-compatible
+                let mut all_exterior_coords = Vec::new();
+                for g in GeometryCollectionTrait::geometries(gc) {
+                    all_exterior_coords.extend(g.exterior_coords_iter());
+                }
+                ExteriorIter::GeometryCollection(Box::new(all_exterior_coords.into_iter()))
+            }
+            GeometryType::Rect(r) => ExteriorIter::Rect(RectIter::new(r)),
+            GeometryType::Triangle(t) => ExteriorIter::Triangle(TriangleIter::new(t)),
+        }
     }
 }
 
 // ┌─────────────────────────┐
 // │ Implementation for Line │
 // └─────────────────────────┘
 
-impl<T: CoordNum> CoordsIter for Line<T> {
-    type Iter<'a>
-        = iter::Chain<iter::Once<Coord<T>>, iter::Once<Coord<T>>>
-    where
-        T: 'a;
-    type ExteriorIter<'a>
-        = Self::Iter<'a>
-    where
-        T: 'a;
-    type Scalar = T;
-
-    fn coords_iter(&self) -> Self::Iter<'_> {
-        iter::once(self.start).chain(iter::once(self.end))
-    }
+/// Helper iterator for Line coordinates
+pub struct LineIter<'a, L: LineTrait + 'a>
```

Comment:
The implementations of the iterators for each geometry type are much more...manual...than the old ones. I couldn't get lifetimes to behave.